### PR TITLE
Enable -Wstring-concatenation and -Wstring-conversion on clang builds

### PR DESCRIPTION
**Pull request by Empact**
The only current violations are a few uses of assert(!"descriptive string").

Seems the benefit of ensuring against a code fault outweighs the benefit of possible additional information in these cases.

> [-Wstring-concatenation](https://clang.llvm.org/docs/DiagnosticsReference.html#wstring-concatenation)
> Diagnostic text:
>
> warning: suspicious concatenation of string literals in an array initialization; did you mean to separate the elements with a comma?
>
> [-Wstring-conversion](https://clang.llvm.org/docs/DiagnosticsReference.html#id805)
> Diagnostic text:
>
> warning: implicit conversion turns string literal into bool: A to B

https://clang.llvm.org/docs/DiagnosticsReference.html#wstring-concatenation

### CODE REVIEW

File: src/net_processing.cpp [hunk 1]
```diff
@@ -2141,17 +2141,15 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
         // Fast-path: in this case it is possible to serve the block directly from disk,
         // as the network format matches the format on disk
         std::vector<uint8_t> block_data;
-        if (!ReadRawBlockFromDisk(block_data, pindex->GetBlockPos(), m_chainparams.MessageStart())) {
-            assert(!"cannot load block from disk");
-        }
+        // fail if cannot load block from disk
+        assert(ReadRawBlockFromDisk(block_data, pindex->GetBlockPos(), m_chainparams.MessageStart()));
```
Comment (theuni) [inline src/net_processing.cpp:2145]: Could you please split this up like this instead?
```c++
auto foo = ReadRawBlockFromDisk(...);
assert(foo);
```
That way the we don't introduce a new side-effect of `NDEBUG=1`

File: src/net_processing.cpp [hunk 2]
```diff
@@ -2141,17 +2141,15 @@ void PeerManagerImpl::ProcessGetBlockData(CNode& pfrom, Peer& peer, const CInv&
...
         m_connman.PushMessage(&pfrom, msgMaker.Make(NetMsgType::BLOCK, Span{block_data}));
         // Don't set pblock as we've sent the block
     } else {
         // Send block from disk
         std::shared_ptr<CBlock> pblockRead = std::make_shared<CBlock>();
-        if (!ReadBlockFromDisk(*pblockRead, pindex, m_chainparams.GetConsensus())) {
-            assert(!"cannot load block from disk");
-        }
+        // fail if cannot load block from disk
+        assert(ReadBlockFromDisk(*pblockRead, pindex, m_chainparams.GetConsensus()));
```
Comment (theuni) [inline src/net_processing.cpp:2152]: Likewise here.

File: src/script/interpreter.cpp [hunk 3]
```diff
@@ -1464,14 +1464,11 @@ const HashWriter HASHER_TAPBRANCH{TaggedHash("TapBranch")};
 
 static bool HandleMissingData(MissingDataBehavior mdb)
 {
-    switch (mdb) {
```
Comment (theuni) [inline src/script/interpreter.cpp:1467]: Imo it'd be cleaner and more readable to just leave the logic here as before, turning the wonky string assertions into comments.

**Review by theuni — CHANGES_REQUESTED** [review]
Concept ACK. This has been open for a while, I'm curious to know if any new violations have snuck in since?

File: src/script/interpreter.cpp (hunk 3 above)
Comment (Empact) [reply src/script/interpreter.cpp:1467]: Done

File: src/net_processing.cpp (hunk 2 above)
Comment (Empact) [reply src/net_processing.cpp]: Done

**Comment by theuni** [comment]
Thanks! Now mind updating to use `Assert()`? :)

**Comment by Empact** [comment]
Rebased, converted assignments to initialization, and moved to `Assert` where possible. I could not apply `Assert(false)` in interpreter.cpp due to errors, but https://github.com/bitcoin/bitcoin/pull/26504 could be applied here once merged.

**Pull request closed without merging**